joomla/joomla-cms
# [4.4] Add information about reasons to Joomla Update when an update was found which doesn't fulfil PHP or DB requirements

Pull Request for Issue #42228 .

### Summary of Changes

This pull request (PR) extends the "Update" object of the updater library to provide additional information about failed checks for PHP and database server requirements and extends the Joomla Update Component's "noupdate" view to show details when there was no suitable update found but one which failed one of these checks.

The change is made in the most b/c way so it uses a new property of the "Update" object and not any existing one, and in case if a suitable update was found there is no change in behaviour.

The language string "COM_JOOMLAUPDATE_NODOWNLOAD_EMPTYSTATE_REASON" for such cases, which was added with PR #42103 , is not modified regarding its meaning or its sprintf parameters. This PR only removes the last part "Please contact your web host to update your server." from it, which will be appended after the new details information again with a new string, and it changes the link to the technical requirements to the page the currently linked site refers to. The latter I was reluctant to do because the currently linked site that linking to another site is just a temporary thing, but the discussion in the referred issue and on social media has shown that people would prefer the link to go directly to the site where currently the requirements can be found. So worst case is that with an outdated translation the removed part of the sting will still be there and the link will be the old one.

### Testing Instructions

You can find a full installation package and an update package of 4.4.2-dev + this PR here: https://artifacts.joomla.org/drone/joomla/joomla-cms/4.4-dev/42489/downloads/72341/ .

If you have a Joomla 4 site which does not meet the PHP or database requirements to update to Joomla 5 you can use that for testing.

If not, you can fake the database or the PHP version or both as described below.

Then make sure that an update is found where the requirements are not met and go to the Joomla Update component.

In addition, check that the Joomla Update Components still works as usual when there is an update found where all requirements are fulfilled.

### How to fake a too low PHP version

Edit file "libraries/src/Updater/Update.php" and search for this line in the "_endElement" method:

```
if (!isset($this->currentUpdate->php_minimum) || version_compare(PHP_VERSION, $this->currentUpdate->php_minimum->_data, '>=')) {
```

In the unmodified file on the 4.4-dev branch it is line 348, in the modified file from this PR it is line 360.

Change the `PHP_VERSION` to e.g. `'7.4.0'`.

In the modified file from this PR you should also do the same change a few lines later in line 367 so you get the faked PHP version also reported in the new details information.

### How to fake a too low database version

Depending on which database driver you use, "MySQLi", "MySQL (PDO)" or "PostgreSQL", modify the "getVersion" method of the particular driver to return a version string with a too small version, e.g. `'5.7.0-0ubuntu0.22.04.1'` for a MySQL, `'10.2.0-0ubuntu0.22.04.1'` for a MariaDB or `'11.0-0ubuntu0.22.04.1'` for PostgreSQL database.
- For "MySQLi" file "libraries/vendor/joomla/database/src/Mysqli/MysqliDriver.php" 
- For "MySQL (PDO)" file "libraries/vendor/joomla/database/src/Mysql/MysqlDriver.php"
- For "PostgreSQL" file "libraries/vendor/joomla/database/src/Pdo/PdoDriver.php"

### Actual result BEFORE applying this Pull Request

See issue #42228 : There are no details on which of the requirements is not met, and the link to the technical requirements goes to a site which refers to another site.

### Expected result AFTER applying this Pull Request

When the PHP version requirement is not met, an additional sentence shows that, telling the current and the required version.

When the database version requirement is not met, an additional sentence shows that, telling the database type and the current and required versions.

### Link to documentations
Please select:
